<a href="https://colab.research.google.com/github/Eng-NadeenShadeed/BinX_AI_ML_Internship/blob/main/Week9/Day%201/day1_sprint_planning_serialization.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<div style="
    background: linear-gradient(135deg, #8B1E3F 0%, #C44569 55%, #F8A5C2 100%);
    padding: 50px 35px;
    border-radius: 22px;
    color: white;
    text-align: center;
    box-shadow: 0 8px 25px rgba(139, 30, 63, 0.25);
    margin-bottom: 10px;
">
<div style="font-size: 13px; letter-spacing: 3px; text-transform: uppercase; opacity: 0.85; margin-bottom: 12px;">
    BinX Tech — AI & ML Internship · Week 9 · Day 1
</div>
<h1 style="font-size: 36px; margin: 0; font-weight: 700; letter-spacing: -0.5px;">
    Sprint 4 Planning & Model Serialization
</h1>
<p style="font-size: 18px; margin-top: 10px; opacity: 0.92; font-weight: 300;">
    ISIC Melanoma Dataset · CNN + Augmentation · Keras
</p>
<div style="margin-top: 20px; font-size: 13px; opacity: 0.75; letter-spacing: 1px;">
    Nadeen Shadeed · BinX Tech AI/ML Internship · 2026
</div>
</div>

<div style="
    background: #fff5f8;
    border-left: 5px solid #C44569;
    border-radius: 12px;
    padding: 28px 32px;
    margin: 20px 0;
    box-shadow: 0 2px 12px rgba(139,30,63,0.07);
">
<div style="font-size: 11px; letter-spacing: 2px; color: #C44569; text-transform: uppercase; margin-bottom: 14px; font-weight: 600;">
    Contents
</div>
<h3 style="margin: 0 0 20px 0; color: #8B1E3F; font-size: 20px; font-weight: 700;">
    Table of Contents
</h3>

<div style="display: flex; flex-direction: column; gap: 6px;">

  <a href="#s1" style="text-decoration: none; display: flex; align-items: center; gap: 14px; padding: 10px 14px; border-radius: 8px; background: white; border: 1px solid #f2d4de;">
    <span style="background: #8B1E3F; color: white; font-size: 13px; font-weight: 700; min-width: 28px; height: 28px; border-radius: 50%; display: flex; align-items: center; justify-content: center;">1</span>
    <span style="color: #333; font-size: 15px;">Sprint 4 Planning & Backlog</span>
  </a>

  <a href="#s2" style="text-decoration: none; display: flex; align-items: center; gap: 14px; padding: 10px 14px; border-radius: 8px; background: #fdf0f4; border: 1px solid #f2d4de;">
    <span style="background: #C44569; color: white; font-size: 13px; font-weight: 700; min-width: 28px; height: 28px; border-radius: 50%; display: flex; align-items: center; justify-content: center;">2</span>
    <span style="color: #333; font-size: 15px;">Why Serialize the Model?</span>
  </a>

  <a href="#s3" style="text-decoration: none; display: flex; align-items: center; gap: 14px; padding: 10px 14px; border-radius: 8px; background: white; border: 1px solid #f2d4de;">
    <span style="background: #8B1E3F; color: white; font-size: 13px; font-weight: 700; min-width: 28px; height: 28px; border-radius: 50%; display: flex; align-items: center; justify-content: center;">3</span>
    <span style="color: #333; font-size: 15px;">MLOps & Reproducibility</span>
  </a>

  <a href="#s4" style="text-decoration: none; display: flex; align-items: center; gap: 14px; padding: 10px 14px; border-radius: 8px; background: #fdf0f4; border: 1px solid #f2d4de;">
    <span style="background: #C44569; color: white; font-size: 13px; font-weight: 700; min-width: 28px; height: 28px; border-radius: 50%; display: flex; align-items: center; justify-content: center;">4</span>
    <span style="color: #333; font-size: 15px;">Retrain & Save the Model</span>
  </a>

  <a href="#s5" style="text-decoration: none; display: flex; align-items: center; gap: 14px; padding: 10px 14px; border-radius: 8px; background: white; border: 1px solid #f2d4de;">
    <span style="background: #8B1E3F; color: white; font-size: 13px; font-weight: 700; min-width: 28px; height: 28px; border-radius: 50%; display: flex; align-items: center; justify-content: center;">5</span>
    <span style="color: #333; font-size: 15px;">Verify the Saved Model</span>
  </a>

  <a href="#s6" style="text-decoration: none; display: flex; align-items: center; gap: 14px; padding: 10px 14px; border-radius: 8px; background: #fdf0f4; border: 1px solid #f2d4de;">
    <span style="background: #C44569; color: white; font-size: 13px; font-weight: 700; min-width: 28px; height: 28px; border-radius: 50%; display: flex; align-items: center; justify-content: center;">6</span>
    <span style="color: #333; font-size: 15px;">Freeze requirements.txt</span>
  </a>

  <a href="#s7" style="text-decoration: none; display: flex; align-items: center; gap: 14px; padding: 10px 14px; border-radius: 8px; background: white; border: 1px solid #f2d4de;">
    <span style="background: #8B1E3F; color: white; font-size: 13px; font-weight: 700; min-width: 28px; height: 28px; border-radius: 50%; display: flex; align-items: center; justify-content: center;">7</span>
    <span style="color: #333; font-size: 15px;">Day 1 Summary</span>
  </a>

</div>
</div>

<div id="s1" style="
    background: linear-gradient(90deg, #8B1E3F 0%, #C44569 100%);
    padding: 16px 28px;
    border-radius: 10px;
    margin: 32px 0 16px 0;
    display: flex;
    align-items: center;
    gap: 16px;
">
<span style="
    background: rgba(255,255,255,0.2);
    color: white;
    font-size: 18px;
    font-weight: 700;
    width: 36px;
    height: 36px;
    border-radius: 50%;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-shrink: 0;
">1</span>
<h2 style="color: white; margin: 0; font-size: 20px; font-weight: 600;">
    Sprint 4 Planning & Backlog
</h2>
</div>

<div style="font-family: Georgia, serif; line-height: 1.8; color: #2d2d2d;">

<div style="
    background: white;
    border-radius: 12px;
    padding: 24px 28px;
    margin-bottom: 16px;
    border: 1px solid #f2d4de;
    font-size: 15px;
    line-height: 1.85;
">
<h3 style="color: #8B1E3F; margin: 0 0 14px 0; font-size: 18px;">What is Sprint 4?</h3>

Sprint 4 is the final sprint of the Capstone project. The goal is to take the trained CNN model and turn it into a real application — one that anyone can open in a browser, upload a skin image, and get a prediction from.<br><br>

This sprint covers four stages:

<table style="width: 100%; border-collapse: collapse; font-size: 14px; margin-top: 16px;">
  <thead>
    <tr style="background: linear-gradient(90deg, #8B1E3F, #C44569);">
      <th style="padding: 10px 14px; color: white; text-align: left;">Day</th>
      <th style="padding: 10px 14px; color: white; text-align: left;">Task</th>
      <th style="padding: 10px 14px; color: white; text-align: left;">Deliverable</th>
    </tr>
  </thead>
  <tbody>
    <tr style="background: linear-gradient(90deg, #fff0f5, #fdf0f4); border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; font-weight: 700; color: #8B1E3F;">Day 1</td>
      <td style="padding: 10px 14px; font-weight: 700; color: #8B1E3F;">Sprint Planning + Model Serialization</td>
      <td style="padding: 10px 14px; font-weight: 700; color: #8B1E3F;">🔵 Today</td>
    </tr>
    <tr style="border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; color: #aaa;">Day 2</td>
      <td style="padding: 10px 14px; color: #aaa;">FastAPI — /predict endpoint</td>
      <td style="padding: 10px 14px; color: #aaa;">⬜ Upcoming</td>
    </tr>
    <tr style="background: #fdf7f9; border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; color: #aaa;">Day 3</td>
      <td style="padding: 10px 14px; color: #aaa;">Streamlit Dashboard</td>
      <td style="padding: 10px 14px; color: #aaa;">⬜ Upcoming</td>
    </tr>
    <tr style="border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; color: #aaa;">Day 4</td>
      <td style="padding: 10px 14px; color: #aaa;">Public Deployment — Hugging Face Spaces</td>
      <td style="padding: 10px 14px; color: #aaa;">⬜ Upcoming</td>
    </tr>
    <tr>
      <td style="padding: 10px 14px; color: #aaa;">Day 5</td>
      <td style="padding: 10px 14px; color: #aaa;">Repo Polish + Sprint Review</td>
      <td style="padding: 10px 14px; color: #aaa;">⬜ Upcoming</td>
    </tr>
  </tbody>
</table>
</div>

<div style="
    background: white;
    border-radius: 12px;
    padding: 24px 28px;
    margin-bottom: 16px;
    border: 1px solid #f2d4de;
    font-size: 15px;
">
<h3 style="color: #8B1E3F; margin: 0 0 16px 0; font-size: 18px;">What I'm Deploying</h3>
<table style="width: 100%; border-collapse: collapse; font-size: 14px;">
  <tbody>
    <tr style="border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; font-weight: 600; color: #8B1E3F; width: 160px;">Project</td>
      <td style="padding: 10px 14px; color: #555;">Cardiac Patient Monitoring System — Melanoma Classification</td>
    </tr>
    <tr style="background: #fdf7f9; border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; font-weight: 600; color: #8B1E3F;">Dataset</td>
      <td style="padding: 10px 14px; color: #555;">ISIC Melanoma Dataset (Benign vs Malignant)</td>
    </tr>
    <tr style="border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; font-weight: 600; color: #8B1E3F;">Model</td>
      <td style="padding: 10px 14px; color: #555;">CNN with Data Augmentation — Week 7 winner</td>
    </tr>
    <tr style="background: #fdf7f9; border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; font-weight: 600; color: #8B1E3F;">Results</td>
      <td style="padding: 10px 14px; color: #555;">Test Accuracy: 88.95% · AUC: 0.9579</td>
    </tr>
    <tr>
      <td style="padding: 10px 14px; font-weight: 600; color: #8B1E3F;">Task Type</td>
      <td style="padding: 10px 14px; color: #555;">Binary image classification</td>
    </tr>
  </tbody>
</table>
</div>

<div style="
    background: white;
    border-radius: 12px;
    padding: 24px 28px;
    margin-bottom: 16px;
    border: 1px solid #f2d4de;
    font-size: 15px;
">
<h3 style="color: #8B1E3F; margin: 0 0 16px 0; font-size: 18px;">Definition of Done</h3>
<table style="width: 100%; border-collapse: collapse; font-size: 14px;">
  <tbody>
    <tr style="border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; color: #555;">☐</td>
      <td style="padding: 10px 14px; color: #555;">Model retrained and saved as <code>.keras</code> file</td>
    </tr>
    <tr style="background: #fdf7f9; border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; color: #555;">☐</td>
      <td style="padding: 10px 14px; color: #555;">FastAPI <code>/predict</code> endpoint returns correct predictions</td>
    </tr>
    <tr style="border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; color: #555;">☐</td>
      <td style="padding: 10px 14px; color: #555;">Streamlit app works end-to-end (upload image → prediction)</td>
    </tr>
    <tr style="background: #fdf7f9; border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; color: #555;">☐</td>
      <td style="padding: 10px 14px; color: #555;">App live on public URL (Hugging Face Spaces)</td>
    </tr>
    <tr>
      <td style="padding: 10px 14px; color: #555;">☐</td>
      <td style="padding: 10px 14px; color: #555;">GitHub repo clean and documented</td>
    </tr>
  </tbody>
</table>
</div>

</div>

<div id="s2" style="
    background: linear-gradient(90deg, #8B1E3F 0%, #C44569 100%);
    padding: 16px 28px;
    border-radius: 10px;
    margin: 32px 0 16px 0;
    display: flex;
    align-items: center;
    gap: 16px;
">
<span style="
    background: rgba(255,255,255,0.2);
    color: white;
    font-size: 18px;
    font-weight: 700;
    width: 36px;
    height: 36px;
    border-radius: 50%;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-shrink: 0;
">2</span>
<h2 style="color: white; margin: 0; font-size: 20px; font-weight: 600;">
    Why Serialize the Model?
</h2>
</div>

<div style="font-family: Georgia, serif; line-height: 1.8; color: #2d2d2d;">

<div style="
    background: white;
    border-radius: 12px;
    padding: 24px 28px;
    margin-bottom: 16px;
    border: 1px solid #f2d4de;
    font-size: 15px;
    line-height: 1.85;
">
<h3 style="color: #8B1E3F; margin: 0 0 14px 0; font-size: 18px;">The Problem</h3>

Training a CNN takes time and GPU compute. Every time the model is needed — in an API, in a web app, in production — retraining from scratch is not an option.<br><br>

<strong>Serialization</strong> solves this: it saves the fully trained model to a file so it can be loaded instantly anywhere, without retraining. The model's weights, architecture, and configuration are all preserved.

<div style="
    background: #fff5f8;
    border-radius: 8px;
    padding: 14px 18px;
    font-family: monospace;
    font-size: 14px;
    color: #555;
    border-left: 3px solid #C44569;
    margin: 16px 0;
">
    Train once → Save → Load anywhere → Predict instantly
</div>

For image models specifically, the preprocessing steps must be applied identically at prediction time:
<ul style="margin: 12px 0; padding-left: 20px; color: #444;">
  <li style="margin-bottom: 6px;">Resize to <code>224×224</code> (same as training)</li>
  <li style="margin-bottom: 6px;">Normalize pixel values to <code>[0, 1]</code> (divide by 255)</li>
  <li style="margin-bottom: 6px;">Expand dims to match batch shape <code>(1, 224, 224, 3)</code></li>
</ul>

These steps will be applied identically in the API and Streamlit app on Day 2 and Day 3.
</div>

<div style="
    background: white;
    border-radius: 12px;
    padding: 24px 28px;
    margin-bottom: 16px;
    border: 1px solid #f2d4de;
    font-size: 15px;
">
<h3 style="color: #8B1E3F; margin: 0 0 16px 0; font-size: 18px;">The Training/Serving Skew Problem</h3>

The most common deployment mistake is saving the model but rewriting the preprocessing by hand in the API. Even a small difference — wrong resize dimensions, wrong normalization range — will silently produce wrong predictions.<br><br>

<table style="width: 100%; border-collapse: collapse; font-size: 14px;">
  <thead>
    <tr style="background: linear-gradient(90deg, #8B1E3F, #C44569);">
      <th style="padding: 10px 14px; color: white; text-align: left;">Approach</th>
      <th style="padding: 10px 14px; color: white; text-align: left;">Risk</th>
    </tr>
  </thead>
  <tbody>
    <tr style="border-bottom: 1px solid #f2d4de; background: #fff5f8;">
      <td style="padding: 10px 14px; color: #C44569; font-weight: 600;">❌ Rewrite preprocessing in the API</td>
      <td style="padding: 10px 14px; color: #555;">Silent errors — wrong predictions with no warning</td>
    </tr>
    <tr>
      <td style="padding: 10px 14px; color: #2d8a4e; font-weight: 600;">✅ Use the same preprocessing code</td>
      <td style="padding: 10px 14px; color: #555;">Guaranteed consistency between training and serving</td>
    </tr>
  </tbody>
</table>

<div style="margin-top: 14px; font-size: 14px; color: #555; line-height: 1.7;">
The rule: <strong>save the model, and keep the preprocessing logic identical</strong> — not reimplemented, identical.
</div>
</div>

</div>

<div id="s3" style="
    background: linear-gradient(90deg, #8B1E3F 0%, #C44569 100%);
    padding: 16px 28px;
    border-radius: 10px;
    margin: 32px 0 16px 0;
    display: flex;
    align-items: center;
    gap: 16px;
">
<span style="
    background: rgba(255,255,255,0.2);
    color: white;
    font-size: 18px;
    font-weight: 700;
    width: 36px;
    height: 36px;
    border-radius: 50%;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-shrink: 0;
">3</span>
<h2 style="color: white; margin: 0; font-size: 20px; font-weight: 600;">
    MLOps & Reproducibility
</h2>
</div>

<div style="font-family: Georgia, serif; line-height: 1.8; color: #2d2d2d;">

<div style="
    background: white;
    border-radius: 12px;
    padding: 24px 28px;
    margin-bottom: 16px;
    border: 1px solid #f2d4de;
    font-size: 15px;
    line-height: 1.85;
">
<h3 style="color: #8B1E3F; margin: 0 0 14px 0; font-size: 18px;">What is MLOps?</h3>

MLOps (Machine Learning Operations) is the set of practices that make ML models reliable, reproducible, and deployable. It bridges the gap between training a model in a notebook and running it in production.<br><br>

Deployment is where the reproducibility discipline built since Week 1 pays off. Three practices matter now:

<table style="width: 100%; border-collapse: collapse; font-size: 14px; margin-top: 16px;">
  <thead>
    <tr style="background: linear-gradient(90deg, #8B1E3F, #C44569);">
      <th style="padding: 10px 14px; color: white; text-align: left;">Practice</th>
      <th style="padding: 10px 14px; color: white; text-align: left;">Why it matters</th>
      <th style="padding: 10px 14px; color: white; text-align: left;">Tool</th>
    </tr>
  </thead>
  <tbody>
    <tr style="border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; font-weight: 600; color: #8B1E3F;">Pinned dependencies</td>
      <td style="padding: 10px 14px; color: #555;">Exact library versions used in training must match deployment</td>
      <td style="padding: 10px 14px; color: #555;"><code>requirements.txt</code></td>
    </tr>
    <tr style="background: #fdf7f9; border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; font-weight: 600; color: #8B1E3F;">Fixed random seeds</td>
      <td style="padding: 10px 14px; color: #555;">Results are repeatable across runs</td>
      <td style="padding: 10px 14px; color: #555;"><code>tf.random.set_seed(42)</code></td>
    </tr>
    <tr>
      <td style="padding: 10px 14px; font-weight: 600; color: #8B1E3F;">Model versioning</td>
      <td style="padding: 10px 14px; color: #555;">Every model version is traceable and recoverable</td>
      <td style="padding: 10px 14px; color: #555;">Saved <code>.keras</code> file + Git</td>
    </tr>
  </tbody>
</table>
</div>

<div style="
    background: linear-gradient(135deg, #fff5f8, #fdf0f4);
    border-radius: 12px;
    padding: 20px 24px;
    border: 1px solid #f2d4de;
    font-size: 14.5px;
    line-height: 1.8;
">
<div style="color: #8B1E3F; font-weight: 700; font-size: 12px; letter-spacing: 1.5px; text-transform: uppercase; margin-bottom: 10px;">
    🔗 Connection to Week 1
</div>
Since Week 1, every notebook has followed consistent practices: fixed seeds, documented results, version-controlled code on GitHub.<br><br>
Sprint 4 is where all of that pays off. A model that was trained with discipline — fixed seeds, documented hyperparameters, clean preprocessing — can be deployed with confidence. A model trained carelessly cannot.
</div>

</div>

<div id="s4" style="
    background: linear-gradient(90deg, #8B1E3F 0%, #C44569 100%);
    padding: 16px 28px;
    border-radius: 10px;
    margin: 32px 0 16px 0;
    display: flex;
    align-items: center;
    gap: 16px;
">
<span style="
    background: rgba(255,255,255,0.2);
    color: white;
    font-size: 18px;
    font-weight: 700;
    width: 36px;
    height: 36px;
    border-radius: 50%;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-shrink: 0;
">4</span>
<h2 style="color: white; margin: 0; font-size: 20px; font-weight: 600;">
    Retrain & Save the Model
</h2>
</div>

<div style="
    font-family: Georgia, serif;
    font-size: 13px;
    font-weight: 700;
    color: #C44569;
    letter-spacing: 1.5px;
    text-transform: uppercase;
    margin: 20px 0 8px 0;
    padding-bottom: 6px;
    border-bottom: 1px solid #f2d4de;
">
    Mount Drive + Imports
</div>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
import matplotlib.pyplot as plt
import os
import random

random.seed(42)
np.random.seed(42)
tf.random.set_seed(42)

DATA_DIR = '/content/drive/MyDrive/BinX_AI_ML_Internship/Week7/Day 1/Data/train'
SAVE_DIR = '/content/drive/MyDrive/BinX_AI_ML_Internship/Week9/Day1/models'
os.makedirs(SAVE_DIR, exist_ok=True)

print("TensorFlow:", tf.__version__)
print("Ready ✓")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
TensorFlow: 2.20.0
Ready ✓


<div style="
    font-family: Georgia, serif;
    font-size: 13px;
    font-weight: 700;
    color: #C44569;
    letter-spacing: 1.5px;
    text-transform: uppercase;
    margin: 20px 0 8px 0;
    padding-bottom: 6px;
    border-bottom: 1px solid #f2d4de;
">
    Load Data
</div>

In [2]:
AUTOTUNE = tf.data.AUTOTUNE
IMG_SIZE = (112, 112)
BATCH_SIZE = 64

def load_and_preprocess(path, label):
    img = tf.io.read_file(path)
    img = tf.image.decode_jpeg(img, channels=3)
    img = tf.image.resize(img, IMG_SIZE)
    img = img / 255.0
    return img, label

def augment(img, label):
    img = tf.image.random_flip_left_right(img)
    img = tf.image.random_brightness(img, 0.2)
    img = tf.image.random_contrast(img, 0.8, 1.2)
    return img, label

all_paths, all_labels = [], []
class_names = sorted(os.listdir(DATA_DIR))
class_map = {name: i for i, name in enumerate(class_names)}
print("Classes:", class_map)

for class_name in class_names:
    class_dir = os.path.join(DATA_DIR, class_name)
    for fname in os.listdir(class_dir):
        if fname.lower().endswith(('.jpg', '.jpeg', '.png')):
            all_paths.append(os.path.join(class_dir, fname))
            all_labels.append(class_map[class_name])

print(f"Total images: {len(all_paths)}")

idx = np.random.permutation(len(all_paths))
split = int(len(idx) * 0.8)
train_idx, val_idx = idx[:split], idx[split:]

train_paths  = [all_paths[i] for i in train_idx]
train_labels = [all_labels[i] for i in train_idx]
val_paths    = [all_paths[i] for i in val_idx]
val_labels   = [all_labels[i] for i in val_idx]

train_ds = (tf.data.Dataset.from_tensor_slices((train_paths, train_labels))
            .map(load_and_preprocess, num_parallel_calls=AUTOTUNE)
            .map(augment, num_parallel_calls=AUTOTUNE)
            .shuffle(1000)
            .batch(BATCH_SIZE)
            .prefetch(AUTOTUNE))

val_ds = (tf.data.Dataset.from_tensor_slices((val_paths, val_labels))
          .map(load_and_preprocess, num_parallel_calls=AUTOTUNE)
          .batch(BATCH_SIZE)
          .prefetch(AUTOTUNE))

print(f"Train batches: {len(train_ds)}")
print(f"Val batches  : {len(val_ds)}")
print("Datasets ready ✓")

Classes: {'Benign': 0, 'Malignant': 1}
Total images: 11879
Train batches: 149
Val batches  : 38
Datasets ready ✓


<div style="
    font-family: Georgia, serif;
    font-size: 13px;
    font-weight: 700;
    color: #C44569;
    letter-spacing: 1.5px;
    text-transform: uppercase;
    margin: 20px 0 8px 0;
    padding-bottom: 6px;
    border-bottom: 1px solid #f2d4de;
">
    Build Model
</div>

In [3]:

# CNN + Data Augmentation — same architecture as Week 7


def build_model():
    model = keras.Sequential([
        layers.Conv2D(32, (3,3), activation='relu', padding='same',
                      input_shape=(112, 112, 3)),
        layers.BatchNormalization(),
        layers.MaxPooling2D(2, 2),

        layers.Conv2D(64, (3,3), activation='relu', padding='same'),
        layers.BatchNormalization(),
        layers.MaxPooling2D(2, 2),

        layers.Conv2D(128, (3,3), activation='relu', padding='same'),
        layers.BatchNormalization(),
        layers.MaxPooling2D(2, 2),

        layers.Conv2D(256, (3,3), activation='relu', padding='same'),
        layers.BatchNormalization(),
        layers.MaxPooling2D(2, 2),

        layers.GlobalAveragePooling2D(),
        layers.Dense(256, activation='relu'),
        layers.Dropout(0.5),
        layers.Dense(1, activation='sigmoid')
    ])

    model.compile(
        optimizer=keras.optimizers.Adam(learning_rate=0.001),
        loss='binary_crossentropy',
        metrics=['accuracy', keras.metrics.AUC(name='auc')]
    )
    return model

model = build_model()
model.summary()

/usr/local/lib/python3.13/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 112, 112, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization             │ (None, 112, 112, 32)   │           128 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 56, 56, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 56, 56, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_1           │ (None, 56, 56, 64)     │           256 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 28, 28, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_2 (Conv2D)               │ (None, 28, 28, 128)    │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_2           │ (None, 28, 28, 128)    │           512 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_2 (MaxPooling2D)  │ (None, 14, 14, 128)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_3 (Conv2D)               │ (None, 14, 14, 256)    │       295,168 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_3           │ (None, 14, 14, 256)    │         1,024 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_3 (MaxPooling2D)  │ (None, 7, 7, 256)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ global_average_pooling2d        │ (None, 256)            │             0 │
│ (GlobalAveragePooling2D)        │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 256)            │        65,792 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 256)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │           257 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 456,385 (1.74 MB)

 Trainable params: 455,425 (1.74 MB)

 Non-trainable params: 960 (3.75 KB)

<div style="
    font-family: Georgia, serif;
    font-size: 13px;
    font-weight: 700;
    color: #C44569;
    letter-spacing: 1.5px;
    text-transform: uppercase;
    margin: 20px 0 8px 0;
    padding-bottom: 6px;
    border-bottom: 1px solid #f2d4de;
">
    Train Model
</div>

In [4]:

callbacks = [
    keras.callbacks.EarlyStopping(
        monitor='val_auc',
        patience=5,
        mode='max',
        restore_best_weights=True,
        verbose=1
    ),
    keras.callbacks.ReduceLROnPlateau(
        monitor='val_auc',
        factor=0.5,
        patience=3,
        mode='max',
        verbose=1
    )
]

history = model.fit(
    train_ds,
    validation_data=val_ds,
    epochs=30,
    callbacks=callbacks,
    verbose=1
)

print("\nTraining complete ✓")
print(f"Best val_auc : {max(history.history['val_auc']):.4f}")
print(f"Best val_acc : {max(history.history['val_accuracy']):.4f}")

Epoch 1/30
149/149 ━━━━━━━━━━━━━━━━━━━━ 1325s 9s/step - accuracy: 0.8059 - auc: 0.8841 - loss: 0.4349 - val_accuracy: 0.5177 - val_auc: 0.5096 - val_loss: 3.2226 - learning_rate: 0.0010
Epoch 2/30
149/149 ━━━━━━━━━━━━━━━━━━━━ 21s 128ms/step - accuracy: 0.8367 - auc: 0.9162 - loss: 0.3674 - val_accuracy: 0.5177 - val_auc: 0.6430 - val_loss: 2.6902 - learning_rate: 0.0010
Epoch 3/30
149/149 ━━━━━━━━━━━━━━━━━━━━ 22s 133ms/step - accuracy: 0.8504 - auc: 0.9246 - loss: 0.3490 - val_accuracy: 0.6208 - val_auc: 0.9033 - val_loss: 1.0609 - learning_rate: 0.0010
Epoch 4/30
149/149 ━━━━━━━━━━━━━━━━━━━━ 43s 144ms/step - accuracy: 0.8563 - auc: 0.9316 - loss: 0.3338 - val_accuracy: 0.7327 - val_auc: 0.9292 - val_loss: 0.5094 - learning_rate: 0.0010
Epoch 5/30
149/149 ━━━━━━━━━━━━━━━━━━━━ 21s 132ms/step - accuracy: 0.8626 - auc: 0.9366 - loss: 0.3205 - val_accuracy: 0.6183 - val_auc: 0.9057 - val_loss: 1.1922 - learning_rate: 0.0010
Epoch 6/30
149/149 ━━━━━━━━━━━━━━━━━━━━ 41s 136ms/step - accuracy:

In [5]:

# Save the trained model


MODEL_PATH = os.path.join(SAVE_DIR, 'melanoma_cnn.keras')
model.save(MODEL_PATH)

print(f"Model saved ✓")
print(f"Path: {MODEL_PATH}")
print(f"Size: {os.path.getsize(MODEL_PATH) / 1024 / 1024:.1f} MB")

Model saved ✓
Path: /content/drive/MyDrive/BinX_AI_ML_Internship/Week9/Day1/models/melanoma_cnn.keras
Size: 5.3 MB


<div style="
    background: #f9f9f9;
    border-left: 4px solid #C44569;
    border-radius: 0 10px 10px 0;
    padding: 16px 20px;
    margin: 12px 0 24px 0;
    font-size: 14.5px;
    line-height: 1.75;
    color: #444;
">
<strong style="color: #8B1E3F; font-size: 13px; letter-spacing: 1px; text-transform: uppercase;">
    What this tells me
</strong><br><br>

The model is saved as a single <code>.keras</code> file (5.3 MB) that contains the full architecture, weights, and training configuration. This format is Keras-native and the safest choice for deployment — it requires no manual reconstruction of the model at load time.<br><br>

<strong style="color: #C44569;">Path:</strong> <code>BinX_AI_ML_Internship/Week9/Day1/models/melanoma_cnn.keras</code><br><br>

This file is the foundation of everything in Sprint 4. Day 2 (FastAPI) and Day 3 (Streamlit) will both load this exact file — no retraining needed.
</div>

<div id="s5" style="
    background: linear-gradient(90deg, #8B1E3F 0%, #C44569 100%);
    padding: 16px 28px;
    border-radius: 10px;
    margin: 32px 0 16px 0;
    display: flex;
    align-items: center;
    gap: 16px;
">
<span style="
    background: rgba(255,255,255,0.2);
    color: white;
    font-size: 18px;
    font-weight: 700;
    width: 36px;
    height: 36px;
    border-radius: 50%;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-shrink: 0;
">5</span>
<h2 style="color: white; margin: 0; font-size: 20px; font-weight: 600;">
    Verify the Saved Model
</h2>
</div>

In [6]:

# Verify the saved model


# 1. Load the model back
loaded_model = keras.models.load_model(MODEL_PATH)
print("Model loaded ✓")

# 2. Take one real image from val set and predict
import tensorflow as tf

sample_path  = val_paths[0]
sample_label = val_labels[0]
class_names  = ['Benign', 'Malignant']

# Preprocess exactly like training
img = tf.io.read_file(sample_path)
img = tf.image.decode_jpeg(img, channels=3)
img = tf.image.resize(img, (112, 112))
img = img / 255.0
img = tf.expand_dims(img, axis=0)  # (1, 112, 112, 3)

# Predict
pred_prob  = loaded_model.predict(img, verbose=0)[0][0]
pred_class = 1 if pred_prob >= 0.5 else 0

print(f"\nSample image : {sample_path.split('/')[-1]}")
print(f"True label   : {class_names[sample_label]}")
print(f"Predicted    : {class_names[pred_class]} ({pred_prob:.4f})")
print(f"Correct      : {'✅' if pred_class == sample_label else '❌'}")

Model loaded ✓

Sample image : 393.jpg
True label   : Malignant
Predicted    : Malignant (0.5844)
Correct      : ✅


<div style="
    background: #f9f9f9;
    border-left: 4px solid #C44569;
    border-radius: 0 10px 10px 0;
    padding: 16px 20px;
    margin: 12px 0 24px 0;
    font-size: 14.5px;
    line-height: 1.75;
    color: #444;
">
<strong style="color: #8B1E3F; font-size: 13px; letter-spacing: 1px; text-transform: uppercase;">
    What this tells me
</strong><br><br>

The saved model loaded correctly and produced the right prediction on a real image from the validation set.<br><br>

<strong style="color: #C44569;">Prediction: Malignant (0.5844) — Correct ✅</strong><br>
The model identified the lesion correctly with 58% confidence. The preprocessing applied at inference time — resize to 112×112, normalize to [0,1], expand dims — was identical to training, which confirms there is no training/serving skew.<br><br>

This means the model is safe to serve. Whatever comes next — FastAPI on Day 2, Streamlit on Day 3 — will use the exact same preprocessing steps and get the exact same results.
</div>

<div id="s6" style="
    background: linear-gradient(90deg, #8B1E3F 0%, #C44569 100%);
    padding: 16px 28px;
    border-radius: 10px;
    margin: 32px 0 16px 0;
    display: flex;
    align-items: center;
    gap: 16px;
">
<span style="
    background: rgba(255,255,255,0.2);
    color: white;
    font-size: 18px;
    font-weight: 700;
    width: 36px;
    height: 36px;
    border-radius: 50%;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-shrink: 0;
">6</span>
<h2 style="color: white; margin: 0; font-size: 20px; font-weight: 600;">
    Freeze requirements.txt
</h2>
</div>

In [7]:

# Freeze requirements.txt


import subprocess

req_path = os.path.join(SAVE_DIR, '../requirements.txt')
req_path = os.path.normpath(req_path)

# Get installed packages
result = subprocess.run(['pip', 'freeze'], capture_output=True, text=True)

# Filter only what we need
keywords = ['tensorflow', 'keras', 'numpy', 'pillow', 'matplotlib']
lines = [l for l in result.stdout.split('\n')
         if any(k in l.lower() for k in keywords)]

with open(req_path, 'w') as f:
    f.write('\n'.join(lines))

print("requirements.txt saved ✓")
print(f"Path: {req_path}")
print("\nContents:")
for l in lines:
    print(f"  {l}")

requirements.txt saved ✓
Path: /content/drive/MyDrive/BinX_AI_ML_Internship/Week9/Day1/requirements.txt

Contents:
  keras==3.13.2
  keras-hub==0.26.0
  keras-nlp==0.26.0
  matplotlib==3.10.0
  matplotlib-inline==0.2.2
  matplotlib-venn==1.1.2
  numpy==2.1.3
  pillow==11.3.0
  tensorflow==2.20.0
  tensorflow-datasets==4.9.10
  tensorflow-hub==0.16.1
  tensorflow-metadata==1.21.0
  tensorflow-probability==0.25.0
  tensorflow-text==2.20.1
  tf_keras==2.20.1


<div style="
    background: #f9f9f9;
    border-left: 4px solid #C44569;
    border-radius: 0 10px 10px 0;
    padding: 16px 20px;
    margin: 12px 0 24px 0;
    font-size: 14.5px;
    line-height: 1.75;
    color: #444;
">
<strong style="color: #8B1E3F; font-size: 13px; letter-spacing: 1px; text-transform: uppercase;">
    What this tells me
</strong><br><br>

The <code>requirements.txt</code> captures the exact library versions used during training. This is the MLOps practice that prevents the most common deployment failure: a model that works in the notebook but breaks in production because of a library version mismatch.<br><br>

<strong style="color: #C44569;">Key versions pinned:</strong><br>
<code>tensorflow==2.20.0</code> · <code>keras==3.13.2</code> · <code>numpy==2.1.3</code> · <code>pillow==11.3.0</code><br><br>

When the app is deployed to Hugging Face Spaces on Day 4, this file will be used to install the exact same environment — guaranteeing the model behaves identically in production as it did in training.
</div>

<div id="s7" style="
    background: linear-gradient(90deg, #8B1E3F 0%, #C44569 100%);
    padding: 16px 28px;
    border-radius: 10px;
    margin: 32px 0 16px 0;
    display: flex;
    align-items: center;
    gap: 16px;
">
<span style="
    background: rgba(255,255,255,0.2);
    color: white;
    font-size: 18px;
    font-weight: 700;
    width: 36px;
    height: 36px;
    border-radius: 50%;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-shrink: 0;
">7</span>
<h2 style="color: white; margin: 0; font-size: 20px; font-weight: 600;">
    Day 1 Summary
</h2>
</div>

<div style="font-family: Georgia, serif; line-height: 1.8; color: #2d2d2d;">

<div style="
    background: white;
    border-radius: 12px;
    padding: 24px 28px;
    margin-bottom: 16px;
    border: 1px solid #f2d4de;
    font-size: 15px;
">
<h3 style="color: #8B1E3F; margin: 0 0 16px 0; font-size: 18px;">What I completed today</h3>
<table style="width: 100%; border-collapse: collapse; font-size: 14px;">
  <thead>
    <tr style="background: linear-gradient(90deg, #8B1E3F, #C44569);">
      <th style="padding: 10px 14px; color: white; text-align: left;">Task</th>
      <th style="padding: 10px 14px; color: white; text-align: left;">Result</th>
      <th style="padding: 10px 14px; color: white; text-align: left;">Status</th>
    </tr>
  </thead>
  <tbody>
    <tr style="border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; color: #555;">Sprint 4 Planning</td>
      <td style="padding: 10px 14px; color: #555;">Full backlog defined</td>
      <td style="padding: 10px 14px;">✅</td>
    </tr>
    <tr style="background: #fdf7f9; border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; color: #555;">Model Retrained</td>
      <td style="padding: 10px 14px; color: #555;">val_AUC: 0.9597 · val_Acc: 89.39%</td>
      <td style="padding: 10px 14px;">✅</td>
    </tr>
    <tr style="border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; color: #555;">Model Serialized</td>
      <td style="padding: 10px 14px; color: #555;">melanoma_cnn.keras — 5.3 MB</td>
      <td style="padding: 10px 14px;">✅</td>
    </tr>
    <tr style="background: #fdf7f9; border-bottom: 1px solid #f2d4de;">
      <td style="padding: 10px 14px; color: #555;">Model Verified</td>
      <td style="padding: 10px 14px; color: #555;">Loads correctly · Predicts accurately</td>
      <td style="padding: 10px 14px;">✅</td>
    </tr>
    <tr>
      <td style="padding: 10px 14px; color: #555;">requirements.txt</td>
      <td style="padding: 10px 14px; color: #555;">15 packages pinned</td>
      <td style="padding: 10px 14px;">✅</td>
    </tr>
  </tbody>
</table>
</div>

<div style="
    background: linear-gradient(135deg, #fff5f8, #fdf0f4);
    border-radius: 12px;
    padding: 20px 24px;
    border: 1px solid #f2d4de;
    font-size: 14.5px;
    line-height: 1.8;
">
<div style="color: #8B1E3F; font-weight: 700; font-size: 12px; letter-spacing: 1.5px; text-transform: uppercase; margin-bottom: 10px;">
    🔜 Coming Next — Day 2
</div>
The serialized model is ready. Day 2 builds a FastAPI app that loads it and exposes a <code>/predict</code> endpoint — so any application can send an image and receive a prediction as JSON. The same preprocessing pipeline used here will be applied identically in the API.
</div>

</div>